lib/os/p4wq.c

```diff
@@ -0,0 +1,209 @@
+ * tolerate equality as meaning "not lessthan"
+ */
+static inline bool item_lessthan(struct k_p4wq_work *a, struct k_p4wq_work *b)
+{
+	if (a->priority > b->priority) {
+		return true;
+	} else if ((a->priority == b->priority) &&
+		   (a->deadline != b->deadline)) {
+		return a->deadline - b->deadline > 0;
+	}
+	return false;
+}
+
+static FUNC_NORETURN void p4wq_loop(void *p0, void *p1, void *p2)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	struct k_p4wq *queue = p0;
+	k_spinlock_key_t k = k_spin_lock(&queue->lock);
+
+	while (true) {
+		struct rbnode *r = rb_get_max(&queue->queue);
+
+		if (r) {
+			struct k_p4wq_work *w
+				= CONTAINER_OF(r, struct k_p4wq_work, rbnode);
+
+			rb_remove(&queue->queue, r);
+			w->thread = _current;
+			sys_dlist_append(&queue->active, &w->dlnode);
+			set_prio(_current, w);
+
+			k_spin_unlock(&queue->lock, k);
+			w->handler(w);
+			k = k_spin_lock(&queue->lock);
+
+			/* Remove from the active list only if it
+			 * wasn't resubmitted already
+			 */
+			if (w->thread == _current) {
+				sys_dlist_remove(&w->dlnode);
+				w->thread = NULL;
+			}
+		} else {
+			z_pend_curr(&queue->lock, k, &queue->waitq, K_FOREVER);
+			k = k_spin_lock(&queue->lock);
+		}
+	}
+}
+
+void k_p4wq_init(struct k_p4wq *queue)
+{
+	memset(queue, 0, sizeof(*queue));
+	z_waitq_init(&queue->waitq);
+	queue->queue.lessthan_fn = rb_lessthan;
+	sys_dlist_init(&queue->active);
+}
+
+void k_p4wq_add_thread(struct k_p4wq *queue, struct k_thread *thread,
+			k_thread_stack_t *stack,
+			size_t stack_size)
+{
+	k_thread_create(thread, stack, stack_size,
+			p4wq_loop, queue, NULL, NULL,
+			K_HIGHEST_THREAD_PRIO, 0, K_NO_WAIT);
+}
+
+static int static_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	Z_STRUCT_SECTION_FOREACH(k_p4wq_initparam, pp) {
+		k_p4wq_init(pp->queue);
+		for (int i = 0; i < pp->num; i++) {
+			uintptr_t ssz = K_THREAD_STACK_LEN(pp->stack_size);
+
+			k_p4wq_add_thread(pp->queue, &pp->threads[i],
+					  &pp->stacks[ssz * i],
+					  pp->stack_size);
+		}
+	}
+	return 0;
+}
+
+/* We spawn a bunch of high priority threads, use the "SMP" initlevel
+ * so they can initialize in parallel instead of serially on the main
+ * CPU.
+ */
+SYS_INIT(static_init, SMP, 99);
+
+void k_p4wq_submit(struct k_p4wq *queue, struct k_p4wq_work *item)
+{
+	k_spinlock_key_t k = k_spin_lock(&queue->lock);
+
+	/* Input is a delta time from now (to match
+	 * k_thread_deadline_set()), but we store and use the absolute
+	 * cycle count.
+	 */
+	item->deadline += k_cycle_get_32();
+
+	/* Resubmission from within handler?  Remove from active list */
+	if (item->thread == _current) {
+		sys_dlist_remove(&item->dlnode);
+		item->thread = NULL;
+	}
+	__ASSERT_NO_MSG(item->thread == NULL);
+
+	rb_insert(&queue->queue, &item->rbnode);
+
+	/* If there were other items already ahead of it in the queue,
+	 * then we don't need to revisit active thread state and can
+	 * return.
+	 */
+	if (rb_get_max(&queue->queue) != &item->rbnode) {
```

Comment:
Sorry, don't understand this either. `rb_get_max()` finds the highest priority work item on the queue, right? If it isn't our new item then we shouldn't preempt it, that's correct. But what if there are more available idle threads / CPUs?

Reply:
If there are idle threads, then the rbtree must be empty.   The tree is there to store backlog work that can't be handled by a synchronous thread yet.

Reply:
got this, thanks!